### PR TITLE
GH-493: Transactional producer factory - set enable.idempotence to tr…

### PR DESCRIPTION
**Pull request by nklmish**
By default idempotency is enabled for a transactional producer , more details #493

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java [hunk 1]
```diff
@@ -129,6 +130,17 @@ public void setPhysicalCloseTimeout(int physicalCloseTimeout) {
...
+		enableIdempotentBehaviour();
+	}
+
+	/**
+	 * When set to 'true', the producer will ensure that exactly one copy of each message is written in the stream.
+	 */
+	private void enableIdempotentBehaviour() {
+		Object previousValue = this.configs.putIfAbsent(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+		if (logger.isDebugEnabled() && previousValue != null && previousValue.equals(false)) {
+			logger.debug(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG +  "is set to " + previousValue + ",may result in duplicate messages");
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java]: Whitespace before `may`, after comma

File: spring-kafka/src/test/java/org/springframework/kafka/core/DefaultKafkaProducerFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Map;
+
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.junit.Test;
+
+/**
+ * @author Nakul Mishra
+ */
+
+public class DefaultKafkaProducerFactoryTest {
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/core/DefaultKafkaProducerFactoryTest.java]: Our pattern is `...Tests` - plural.
More over I think we can move your two tests into the existing `KafkaTemplateTransactionTests`.

**Comment by nklmish** [comment]
Thank you for the review! PR updated.

File: spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java [hunk 3]
```diff
@@ -129,6 +130,17 @@ public void setPhysicalCloseTimeout(int physicalCloseTimeout) {
...
+		enableIdempotentBehaviour();
+	}
+
+	/**
+	 * When set to 'true', the producer will ensure that exactly one copy of each message is written in the stream.
+	 */
+	private void enableIdempotentBehaviour() {
+		Object previousValue = this.configs.putIfAbsent(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+		if (logger.isDebugEnabled() && previousValue != null && previousValue.equals(false)) {
+			logger.debug(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG +  "is set to " + previousValue + ", may result in duplicate messages");
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java]: The `ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG` is resolved to the `enable.idempotence` string and then you have an `is` word. But there is no whitespace between them.

I would wrap `ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG` to the single quotes and add `The ` in the beginning of sentence.

File: spring-kafka/src/test/java/org/springframework/kafka/core/DefaultKafkaProducerFactoryTests.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Map;
+
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.junit.Test;
+
+/**
+ * @author Nakul Mishra
+ */
+
+public class DefaultKafkaProducerFactoryTests {
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/core/DefaultKafkaProducerFactoryTests.java]: You didn't understand me: we don't need this class at all.
Your `testDefaultProducerIdempotentConfig()` and `testOverrideProducerIdempotentConfig()` can simply be moved to the existing `KafkaTemplateTransactionTests` class.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/ebd01beba8c64185c1481241391cde7ab9f9e37e.

Thank you for contribution.
Please, consider to configure your Git client for your real name, so it will appear in commits properly.

**Pull request closed without merging**